File: src/validate_credit_card.py
import re

def validate_credit_card(card_number):
    # Remove all non-digit characters
    sanitized = re.sub(r'\D', '', card_number)

    # Check if the card number is valid using Luhn Algorithm
    if not luhn_check(sanitized):
        return {'valid': False, 'bandeira': None}

    # Determine the card issuer (bandeira)
    bandeira = get_card_issuer(sanitized)

    return {'valid': True, 'bandeira': bandeira}

def luhn_check(card_number):
    sum = 0
    should_double = False

    # Loop through the card number digits in reverse order
    for digit in reversed(card_number):
        digit = int(digit)

        if should_double:
            digit *= 2
            if digit > 9:
                digit -= 9

        sum += digit
        should_double = not should_double

    return sum % 10 == 0

def get_card_issuer(card_number):
    patterns = {
        'visa': r'^4[0-9]{12}(?:[0-9]{3})?$',
        'mastercard': r'^5[1-5][0-9]{14}$',
        'amex': r'^3[47][0-9]{13}$',
        'discover': r'^6(?:011|5[0-9]{2})[0-9]{12}$',
        'diners': r'^3(?:0[0-5]|[68][0-9])[0-9]{11}$',
        'jcb': r'^(?:2131|1800|35\d{3})\d{11}$',
        'elo': r'^((636368)|(438935)|(504175)|(451416)|(636297)|(5067)|(4576)|(4011))\d+$',
        'hipercard': r'^(606282\d{10}(\d{3})?)|(3841\d{15})$',
        'aura': r'^50[0-9]{14,17}$'
    }

    for issuer, pattern in patterns.items():
        if re.match(pattern, card_number):
            return issuer

    return 'unknown'

# Example usage for Visa
visa_card_number = '4111111111111111'
visa_result = validate_credit_card(visa_card_number)
print(f"Card Number: {visa_card_number}, Valid: {visa_result['valid']}, Bandeira: {visa_result['bandeira']}") # { valid: True, bandeira: 'visa' }

# Example usage for MasterCard
master_card_number = '5277959558870483'
master_card_result = validate_credit_card(master_card_number)
print(f"Card Number: {master_card_number}, Valid: {master_card_result['valid']}, Bandeira: {master_card_result['bandeira']}") # { valid: True, bandeira: 'mastercard' }

# Example usage for American Express
amex_card_number = '378282246310005'
amex_result = validate_credit_card(amex_card_number)
print(f"Card Number: {amex_card_number}, Valid: {amex_result['valid']}, Bandeira: {amex_result['bandeira']}") # { valid: True, bandeira: 'amex' }

# Example usage for Discover
discover_card_number = '6011111111111117'
discover_result = validate_credit_card(discover_card_number)
print(f"Card Number: {discover_card_number}, Valid: {discover_result['valid']}, Bandeira: {discover_result['bandeira']}") # { valid: True, bandeira: 'discover' }

# Example usage for Diners Club
diners_card_number = '30569309025904'
diners_result = validate_credit_card(diners_card_number)
print(f"Card Number: {diners_card_number}, Valid: {diners_result['valid']}, Bandeira: {diners_result['bandeira']}") # { valid: True, bandeira: 'diners' }

# Example usage for JCB
jcb_card_number = '3530111333300000'
jcb_result = validate_credit_card(jcb_card_number)
print(f"Card Number: {jcb_card_number}, Valid: {jcb_result['valid']}, Bandeira: {jcb_result['bandeira']}") # { valid: True, bandeira: 'jcb' }

# Example usage for Elo
elo_card_number = '6363680000000000'
elo_result = validate_credit_card(elo_card_number)
print(f"Card Number: {elo_card_number}, Valid: {elo_result['valid']}, Bandeira: {elo_result['bandeira']}") # { valid: True, bandeira: 'elo' }

# Example usage for Hipercard
hipercard_card_number = '6062825624254001'
hipercard_result = validate_credit_card(hipercard_card_number)
print(f"Card Number: {hipercard_card_number}, Valid: {hipercard_result['valid']}, Bandeira: {hipercard_result['bandeira']}") # { valid: True, bandeira: 'hipercard' }

# Example usage for Aura
aura_card_number = '5078601870000127980'
aura_result = validate_credit_card(aura_card_number)
print(f"Card Number: {aura_card_number}, Valid: {aura_result['valid']}, Bandeira: {aura_result['bandeira']}") # { valid: True, bandeira: 'aura' }

File: src/ValidateCreditCard.java
import java.util.regex.Pattern;

public class ValidateCreditCard {

    public static void main(String[] args) {
        // Example usage for Visa
        String visaCardNumber = "4111111111111111";
        Result visaResult = validateCreditCard(visaCardNumber);
        System.out.println("Card Number: " + visaCardNumber + ", Valid: " + visaResult.valid + ", Bandeira: "
                + visaResult.bandeira);

        // Example usage for MasterCard
        String masterCardNumber = "5277959558870483";
        Result masterCardResult = validateCreditCard(masterCardNumber);
        System.out.println("Card Number: " + masterCardNumber + ", Valid: " + masterCardResult.valid + ", Bandeira: "
                + masterCardResult.bandeira);

        // Example usage for American Express
        String amexCardNumber = "378282246310005";
        Result amexResult = validateCreditCard(amexCardNumber);
        System.out.println("Card Number: " + amexCardNumber + ", Valid: " + amexResult.valid + ", Bandeira: "
                + amexResult.bandeira);

        // Example usage for Discover
        String discoverCardNumber = "6011111111111117";
        Result discoverResult = validateCreditCard(discoverCardNumber);
        System.out.println("Card Number: " + discoverCardNumber + ", Valid: " + discoverResult.valid + ", Bandeira: "
                + discoverResult.bandeira);

        // Example usage for Diners Club
        String dinersCardNumber = "30569309025904";
        Result dinersResult = validateCreditCard(dinersCardNumber);
        System.out.println("Card Number: " + dinersCardNumber + ", Valid: " + dinersResult.valid + ", Bandeira: "
                + dinersResult.bandeira);

        // Example usage for JCB
        String jcbCardNumber = "3530111333300000";
        Result jcbResult = validateCreditCard(jcbCardNumber);
        System.out.println(
                "Card Number: " + jcbCardNumber + ", Valid: " + jcbResult.valid + ", Bandeira: " + jcbResult.bandeira);

        // Example usage for Elo
        String eloCardNumber = "6363680000000000";
        Result eloResult = validateCreditCard(eloCardNumber);
        System.out.println(
                "Card Number: " + eloCardNumber + ", Valid: " + eloResult.valid + ", Bandeira: " + eloResult.bandeira);

        // Example usage for Hipercard
        String hipercardCardNumber = "6062825624254001";
        Result hipercardResult = validateCreditCard(hipercardCardNumber);
        System.out.println("Card Number: " + hipercardCardNumber + ", Valid: " + hipercardResult.valid + ", Bandeira: "
                + hipercardResult.bandeira);

        // Example usage for Aura
        String auraCardNumber = "5078601870000127980";
        Result auraResult = validateCreditCard(auraCardNumber);
        System.out.println("Card Number: " + auraCardNumber + ", Valid: " + auraResult.valid + ", Bandeira: "
                + auraResult.bandeira);
    }

    public static Result validateCreditCard(String cardNumber) {
        // Remove all non-digit characters
        String sanitized = cardNumber.replaceAll("\\D", "");

        // Check if the card number is valid using Luhn Algorithm
        if (!luhnCheck(sanitized)) {
            return new Result(false, null);
        }

        // Determine the card issuer (bandeira)
        String bandeira = getCardIssuer(sanitized);

        return new Result(true, bandeira);
    }

    public static boolean luhnCheck(String cardNumber) {
        int sum = 0;
        boolean shouldDouble = false;

        // Loop through the card number digits in reverse order
        for (int i = cardNumber.length() - 1; i >= 0; i--) {
            int digit = Character.getNumericValue(cardNumber.charAt(i));

            if (shouldDouble) {
                digit *= 2;
                if (digit > 9) {
                    digit -= 9;
                }
            }

            sum += digit;
            shouldDouble = !shouldDouble;
        }

        return sum % 10 == 0;
    }

    public static String getCardIssuer(String cardNumber) {
        String[] patterns = {
                "^4[0-9]{12}(?:[0-9]{3})?$", // Visa
                "^5[1-5][0-9]{14}$", // MasterCard
                "^3[47][0-9]{13}$", // American Express
                "^6(?:011|5[0-9]{2})[0-9]{12}$", // Discover
                "^3(?:0[0-5]|[68][0-9])[0-9]{11}$", // Diners Club
                "^(?:2131|1800|35\\d{3})\\d{11}$", // JCB
                "^((636368)|(438935)|(504175)|(451416)|(636297)|(5067)|(4576)|(4011))\\d+$", // Elo
                "^(606282\\d{10}(\\d{3})?)|(3841\\d{15})$", // Hipercard
                "^50[0-9]{14,17}$" // Aura
        };

        String[] issuers = {
                "visa", "mastercard", "amex", "discover", "diners", "jcb", "elo", "hipercard", "aura"
        };

        for (int i = 0; i < patterns.length; i++) {
            if (Pattern.matches(patterns[i], cardNumber)) {
                return issuers[i];
            }
        }

        return "unknown";
    }

    static class Result {
        boolean valid;
        String bandeira;

        Result(boolean valid, String bandeira) {
            this.valid = valid;
            this.bandeira = bandeira;
        }
    }
}